purldb-toolkit/src/purldb_toolkit/purlcli.py

```diff
@@ -194,38 +228,27 @@ def construct_headers(
         if not purl:
             continue
 
-        # `metadata` warnings:
-        metadata_purl = check_metadata_purl(purl)
-
-        if command_name == "metadata" and metadata_purl == "not_valid":
-            warnings.append(f"'{purl}' not valid")
-            continue
-
-        if command_name == "metadata" and metadata_purl == "valid_but_not_supported":
-            warnings.append(f"'{purl}' not supported with `metadata` command")
-            continue
-
-        if command_name == "metadata" and metadata_purl == "not_in_upstream_repo":
-            warnings.append(f"'{purl}' does not exist in the upstream repo")
-            continue
-
-        # `urls` warnings:
-        urls_purl = check_urls_purl(purl)
-
-        if command_name == "urls" and urls_purl == "not_valid":
-            warnings.append(f"'{purl}' not valid")
-            continue
+        warning_text = {
+            "error_fetching_purl": f"'error fetching {purl}'",
+            "validation_error": f"'{purl}' encountered a validation error",
+            "not_valid": f"'{purl}' not valid",
+            "valid_but_not_supported": f"'{purl}' not supported with `{command_name}` command",
+            "valid_but_not_fully_supported": f"'{purl}' not fully supported with `urls` command",
+            "not_in_upstream_repo": f"'{purl}' does not exist in the upstream repo",
+        }
 
-        if command_name == "urls" and urls_purl == "valid_but_not_supported":
-            warnings.append(f"'{purl}' not supported with `urls` command")
-            continue
+        if command_name in ["metadata", "urls", "validate", "versions"]:
+            purl_warning = purl_warnings.get(purl, None)
+            if purl_warning:
+                warnings.append(warning_text[purl_warning])
+                print(warning_text[purl_warning])
+                continue
 
-        if command_name == "urls" and urls_purl == "valid_but_not_fully_supported":
-            warnings.append(f"'{purl}' not fully supported with `urls` command")
-
-        if command_name == "urls" and urls_purl == "not_in_upstream_repo":
-            warnings.append(f"'{purl}' does not exist in the upstream repo")
-            continue
+    log_file = Path("purldb-toolkit/src/purldb_toolkit/app.log")
```

Comment:
I think the log file should be saved in the users home directory, or some other canonical place for log files.

Reply:
@JonoYang I'm encountering syntax errors trying to save the `app.log` in the user's home directory.  Variations on 

```
filename = os.path.join(os.path.expanduser("~"), 'app.log')
```

keep returning 

```
    filename = os.path.join(os.path.expanduser("~"), 'app.log')
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: invalid syntax. Perhaps you forgot a comma?
```

I've explored a number of stackoverflow and similar hints (nothing re saving but examples of finding the user's home and expanding, which should be applicable) but no luck yet.  Do you know what the proper syntax is?

Reply:
@johnmhoran I'm not sure what the issue is. I'm able to run `filename = os.path.join(os.path.expanduser("~"), 'app.log')` in the Python shell. Maybe you can try `filename = Path.home() / 'app.log'`

Reply:
Thanks @JonoYang -- I got it working in the users home directory a while ago -- user error.

```
        logging.basicConfig(
            filename=os.path.join(os.path.expanduser("~"), "app.log"),
            level=logging.ERROR,
            format="%(levelname)s - %(message)s",
            filemode="w",
        )

        logger.error(f"validate_purl(): json.decoder.JSONDecodeError for '{purl}': {e}")
```